### ISSUE TITLE
Parallelizing tokenization?  (enhancement)

### ISSUE THREAD

**Issue by drscotthawley** [Natooz/MidiTok#102]
Hi, great code!  Congratulations.  
I'm trying MidiTok and am running the "Full_Example_HuggingFace_GPT2_Transformer" example. 
I noticed that the "Tokenizing MIDIs (Maestro_tokens_no_bpe)" step is going to take 30 minutes on my machine, and when I checked `top` it seems that only one processor is being used.   

Perhaps this is a naive question but: I wonder if tokenizing a large number of files like Maestro could be (partially) parallelized,  similar to how reading in a bunch of midi files via `pretty_midi` (which can take a long time on one processor) can be parallelized via multiprocessing & `pool.map` or `tqdm.contrib.concurrent.process_map` such as...

```python
def process_one_midi_file(filename):
    pm = pretty_midi.PrettyMIDI(midi_file)
    midi_data = pm... # do_stuff.... 
    return midi_data 

def read_files(filenames): 
    midi_data_list = process_map(process_one_midi_file, filenames, max_workers=mp.cpu_count(), chunksize=1)
    return midi_data_list
``` 

Does this seem like such a method for speeding up the tokenization of a large number of files would be feasible?  

Presumably one will still need all the data in one place to analyze it and develop tokens, but perhaps still the initial reading of the files themselves could be parallelized such as via the above code.  
 I have some experience with multiprocessing operations and could try to help, but I'm not familiar with how the REMI tokenizer works.

**Comment by Natooz** [comment]
Hello @drscotthawley ,
Thank you for you kind comments and recommendations! 😃

Indeed, as of right now everything is performed on one thread, and multithreading is definitely something that could speed-up tokenizing a large number of files.
This was actually something that was on my roadmap soon after first releasing the library. I did an implementation (but cannot remember how), but it unfortunately resulted in a longer tokenization time on a Skylake i7 (8 threads). :/
Now I probably did something wrong, but didn't have the time to dig further so just put the idea aside. I don't have much experience into multithreading.

I you are up to it, I'll be glad to provide any help if necessary! Otherwise I could probably get back into it in a few weeks, mid / late December.

First, maybe we should think about which steps could be parallelized. The workflow of MidiTok could be summarized as:

1. Load the MIDI, with `miditoolkit` (same as `pretty_midi` but works in tick instead of seconds). This is a slow step that might be parallelized indeed;
2. Preprocess the MIDI (align notes onset / offset notes, velocity values...);
3. Gather the "global events" (tempo, time signature);
4. Gather the track events of each tracks (notes, sustain pedal...);
5. [Tokenizer dependent] deduce the time events / tokens (`TimeShift`, `Bar`, `Position`) and others tokenizer-dependent, and returns the tokens;

I think step 1 alone parallelized would lead to good speed-ups. Now steps 2 to 4 are all performed by the base `MIDITokenizer` class which is "tokenizer-agnostic", and could maybe be parallelized as well.

PS: ultimately, the best solution would be to load the MIDI directly in C++ or Rust, and maybe even some of the steps above. But this is another topic and really not programed here. It would mean getting rid of `miditoolkit` which relies on `mido`. Actually, I think the authors of `pretty_midi` also had in their roadmap the goal to switch from `mido` to a C++ backend.

**Comment by Yikai-Liao** [comment]
> Hello @drscotthawley ,
> Thank you for you kind comments and recommendations! 😃
> 
> Indeed, as of right now everything is performed on one thread, and multithreading is definitely something that could speed-up tokenizing a large number of files.
> This was actually something that was on my roadmap soon after first releasing the library. I did an implementation (but cannot remember how), but it unfortunately resulted in a longer tokenization time on a Skylake i7 (8 threads). :/
> Now I probably did something wrong, but didn't have the time to dig further so just put the idea aside. I don't have much experience into multithreading.
> 
> I you are up to it, I'll be glad to provide any help if necessary! Otherwise I could probably get back into it in a few weeks, mid / late December.
> 
> First, maybe we should think about which steps could be parallelized. The workflow of MidiTok could be summarized as:
> 
> 1. Load the MIDI, with `miditoolkit` (same as `pretty_midi` but works in tick instead of seconds). This is a slow step that might be parallelized indeed;
> 2. Preprocess the MIDI (align notes onset / offset notes, velocity values...);
> 3. Gather the "global events" (tempo, time signature);
> 4. Gather the track events of each tracks (notes, sustain pedal...);
> 5. [Tokenizer dependent] deduce the time events / tokens (`TimeShift`, `Bar`, `Position`) and others tokenizer-dependent, and returns the tokens;
> 
> I think step 1 alone parallelized would lead to good speed-ups. Now steps 2 to 4 are all performed by the base `MIDITokenizer` class which is "tokenizer-agnostic", and could maybe be parallelized as well.
> 
> PS: ultimately, the best solution would be to load the MIDI directly in C++ or Rust, and maybe even some of the steps above. But this is another topic and really not programed here. It would mean getting rid of `miditoolkit` which relies on `mido`. Actually, I think the authors of `pretty_midi` also had in their roadmap the goal to switch from `mido` to a C++ backend.

hi， I have writen a new midi parsing backend in c++, and opened a new issue

**Comment by Yikai-Liao** [comment]
> @Yikai-Liao does this use multiple threads?

Well, we don't use multiple threads. Because for a normal midi file, the libaray gets millisecond level parsing speed. Therefore, using multiple threads will contribute to a large scheduling overhead.

It will be more effecient to use multiple thread to process different midi files in parallel. And I recommend you try the libarary yourself, as it is available on pypi.

```python
# pip install symusic
from symusic import Score
Score("path to your midi file")
```

Here is the benchmark:

* test using [mahler.mid](https://github.com/lzqlzzq/minimidi/blob/main/example/mahler.mid) from minimidi/example in colab. Note that it is quite a large midi file.
* `mido` is writen in pure python, and only parses midi files to `event level`
* `pretty_midi` and `miditoolkit` is based on `mido`, and parse midi files to `note level`

|libarary| time |
| -- | -- |
|[**symusic**](https://github.com/Yikai-Liao/symusic)|21.8 ms ± 11.7 ms|
|[MIDI.jl](https://github.com/JuliaMusic/MIDI.jl)|128.024 ms|
|[mido](https://github.com/mido/mido)|5.68 s ± 2 s|
|[pretty_midi](https://github.com/craffel/pretty-midi)|5.59 s ± 844 ms|
|[miditoolkit](https://github.com/YatingMusic/miditoolkit)|6.27 s ± 1.79 s|
|[music21](https://github.com/cuthbertLab/music21)|8.59 s ± 1.2 s|
